src/util/unesc.js

```diff
@@ -1,11 +1,19 @@
-const HEX_ESC = /\\(?:([0-9a-fA-F]{6})|([0-9a-fA-F]{1,5})(?: |(?![0-9a-fA-F])))/g;
-const OTHER_ESC = /\\(.)/g;
+const whitespace = '[\\x20\\t\\r\\n\\f]';
+const unescapeRegExp = new RegExp('\\\\([\\da-f]{1,6}' + whitespace + '?|(' + whitespace + ')|.)', 'ig');
+
 export default function unesc (str) {
-    str = str.replace(HEX_ESC, (_, hex1, hex2) => {
-        let hex = hex1 || hex2;
-        let code = parseInt(hex, 16);
-        return String.fromCharCode(code);
+    return str.replace(unescapeRegExp, (_, escaped, escapedWhitespace) => {
+        const high = '0x' + escaped - 0x10000;
+
+        // NaN means non-codepoint
+        // Workaround erroneous numeric interpretation of +"0x"
+        // eslint-disable-next-line no-self-compare
+        return high !== high || escapedWhitespace
+            ? escaped
+            : high < 0
+                ? // BMP codepoint
+                String.fromCharCode(high + 0x10000)
+                : // Supplemental Plane codepoint (surrogate pair)
+                String.fromCharCode((high >> 10) | 0xd800, (high & 0x3ff) | 0xdc00);
```

Comment:
Jquery port https://github.com/jquery/sizzle/blob/master/src/sizzle.js#L152 (very stable and testing in many env/browsers)